### DISCUSSION TITLE
form.setValues设置表单值的合并策略想法

### DISCUSSION THREAD

**Discussion by li9269391** [alibaba/formily#1637] [Ideas]
现有策略 
```type IFormMergeStrategy = 'overwrite' | 'merge' | 'deepMerge' | 'shallowMerge'```

在表单编辑场景，通常会先从接口获取数据设置到表单中，然后进行编辑，接口往往会多出一些额外的字段，如创建时间等。
那么在提交表单值时，无意间多了表单额外的 value。有没有考虑一种合并策略，以原 Form 中的 field name 为基准合并。

举例：
```
原表单值  {a: 1, b: 2}
执行 form.setValues( {a: 2, c: 1} )，期望表单结果为 {a: 2, b: 2}
```

**Comment by liuweiGL** [comment]
你应该自己处理一下，表单没法知道你哪个字段是需要的。

**Reply by li9269391** [reply]
也可以说是知道的，即定义过的 field，没定义过的是不需要的

**Comment by janryWang** [comment]
值合并的边界条件太多了，你这个属于纯业务逻辑

**Reply by li9269391** [reply]
边界太多是同意的，是否纯业务这个观点我觉得是保留意见

**Reply by janryWang** [reply]
本来就是纯业务逻辑啊，因为不具备通用性，你咋就知道定义过的字段要提交，没定义的就过滤？这个规则是基于什么思路来做的？

**Reply by li9269391** [reply]
用户使用时，主动选择策略，那就是已知场景呀

**Reply by li9269391** [reply]
那什么时候用 overwrite 是一个道理吧

**Reply by darkmice** [reply]
这种应该后端去考虑哪些值需要，哪些不需要，前端只要保证不少值，多传值总归不会有问题

**Reply by janryWang** [reply]
> 那什么时候用 overwrite 是一个道理吧

覆盖是一句话能说清楚的，覆盖就是覆盖，你那个策略，要不你帮忙命名，看看命啥名字合适，一句话都不能说清楚的，那就不是个优雅设计

**Reply by li9269391** [reply]
> > 那什么时候用 overwrite 是一个道理吧
> 
> 覆盖是一句话能说清楚的，覆盖就是覆盖，你那个策略，要不你帮忙命名，看看命啥名字合适，一句话都不能说清楚的，那就不是个优雅设计

到点上了，这也是我想说的，我先想想

**Reply by li9269391** [reply]
> 这种应该后端去考虑哪些值需要，哪些不需要，前端只要保证不少值，多传值总归不会有问题

你说的是对的，但这个有不同场景做不到，所以还是回到讨论合并策略，因为我好像见过哪个知名工具库有这么设计合并值策略，我先想想

**Comment by janryWang** [comment]
你说的这种合并策略存在几个问题：
1. 说不清楚依据，已定义字段和未定义字段，保留和不保留，以什么为准？标准在哪里
2. 会增加系统复杂度，因为与 'deepMerge' | 'shallowMerge' 两个策略会出现组合关系，大概会增加4种组合关系才是完备的
3. 不优雅

**Reply by li9269391** [reply]
了解，暂时写了个 filter 方法，在设置之前先洗一下数据再扔给 formily 了

**Reply by zhousufa** [reply]
我也遇到这个问题，多传给formily的数据，就是说不在schema里的key，希望formily能帮忙过滤掉。因为schema通过designable进行了修改，但是已经提交的数据不知道schema进行了修改，所以传给formily的数据有可能多了

**Reply by wwyx778** [reply]
> 你说的这种合并策略存在几个问题：
> 
> 1. 说不清楚依据，已定义字段和未定义字段，保留和不保留，以什么为准？标准在哪里
> 2. 会增加系统复杂度，因为与 'deepMerge' | 'shallowMerge' 两个策略会出现组合关系，大概会增加4种组合关系才是完备的
> 3. 不优雅

我也认为 Field 和 Value 应该有一个一一对应的"注册"关系，即当我在 setValue 时，如果某一个字段在 Graph 上不存在，这个值就应该被丢弃，不再应该被提交。antd 也符合这种逻辑。

我现在遇到的问题就是原表单({ a: 1, b: 1 }) Schema 升级了({ a: 1 })，删除了其中一个字段，但是数据还是旧的数据，我提交时就没法把 b 字段删掉。这种场景下 Schema 的更新是没有代码改动的，如果要手动把 b 过滤掉，我不仅要改代码，甚至还要把 b 字段名写死来过滤，这显然不符合预期。

**Reply by li9269391** [reply]
> 题就是原表单({ a: 1, b: 1 }) Schema 升级了({ a: 1 })，删除了其中一个字段，但是数据还是旧的数据，我提交时就没法把 b 字段删掉。这种场景下 Sc

@wwyx778 你这个问题，可以尝试将form实例与schema一起管理，即schema有变更，form副作用重新实例化

**Reply by wwyx778** [reply]
> > 题就是原表单({ a: 1, b: 1 }) Schema 升级了({ a: 1 })，删除了其中一个字段，但是数据还是旧的数据，我提交时就没法把 b 字段删掉。这种场景下 Sc
> 
> @wwyx778 你这个问题，可以尝试将form实例与schema一起管理，即schema有变更，form副作用重新实例化

@li9269391 无法达到预期。重新 createForm 和数据更新是两码事。
```
const form = React.useMemo(() => createForm({ initialValues: initValue }), [initValue])
```
initValue 更新，会重新实例化 form，但是这并不会影响到我在实例化过程中传给 form 的初始值。这里 initValue 从服务端获取，始终包含我所不需要的 b 字段。

**Reply by atheist1** [reply]
同上，比如我从A schema更换成B schema，期望不重复填充存在相同key的部分，过滤掉在B schema中不存在的地方。
采取的方式也是类似的，重新创建实例，然后设置initialvalues。
我可以接受业务方自己做这个filter动作，但是A、B schema都是通过服务端储存的，现在似乎没办法通过form实例获取到哪些字段存在哪些字段不存在？
